modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/linktoolbar/plugin.js

```diff
@@ -66,197 +106,156 @@
-				const imageWidget = this.editor.widgets.selected[0];
+					if (linkTarget) {
+						const hasTarget = linkTarget.value !== 'Default';
 
-				if (!imageWidget || imageWidget.name !== 'image') {
-					return;
-				}
+						if (hasTarget) {
+							attributes.target = linkTarget.value;
+						}
+					}
 
-				const link = imageWidget.wrapper.findOne('a');
+					if (editor.widgets.selected[0]) {
+						const imageWidget = editor.widgets.selected[0];
 
-				if (link) {
-					link.setAttributes(attributes);
-				}
-				else {
-					const newData = Object.assign(imageWidget.data, {
-						link: {
-							url: linkInput.value,
-						},
-					});
+						if (!imageWidget || imageWidget.name !== 'image') {
+							return;
+						}
 
-					imageWidget.shiftState({
-						deflate() {},
-						element: imageWidget.element,
-						inflate() {
-							const link = imageWidget.wrapper.findOne('a');
+						const link = imageWidget.wrapper.findOne('a');
 
+						if (link) {
 							link.setAttributes(attributes);
-						},
-						newData,
-						oldData: imageWidget.oldData,
-						widget: imageWidget,
-					});
-				}
-
-				this._toolbar.hide();
-
-				imageWidget.focus();
-				imageWidget.element.focus();
-			}
-			else {
-				const startElement = selection.getStartElement();
-
-				let linkElement;
-
-				if (startElement.getName() === 'a') {
-					linkElement = startElement;
+						}
+						else {
+							const newData = Object.assign(imageWidget.data, {
+								link: {
+									url: linkInput.value,
+								},
+							});
+
+							imageWidget.shiftState({
+								deflate() {},
+								element: imageWidget.element,
+								inflate() {
+									const link = imageWidget.wrapper.findOne(
+										'a'
+									);
+
+									link.setAttributes(attributes);
+								},
+								newData,
+								oldData: imageWidget.oldData,
+								widget: imageWidget,
+							});
+						}
+
+						instance._toolbar.hide();
+
+						imageWidget.focus();
+						imageWidget.element.focus();
+					}
+					else {
+						const startElement = selection.getStartElement();
 
-					linkElement.setAttributes(attributes);
+						let linkElement;
 
-					linkInput.clear();
+						if (startElement.getName() === 'a') {
+							linkElement = startElement;
 
-					this._toolbar.hide();
-				}
-				else if (selection.getSelectedText()) {
-					selection = this.editor.getSelection();
+							linkElement.setAttributes(attributes);
 
-					const selectedText = selection.getSelectedText();
+							linkInput.clear();
 
-					const ranges = selection.getRanges();
+							instance._toolbar.hide();
+						}
+						else if (selection.getSelectedText()) {
+							selection = editor.getSelection();
 
-					const range = ranges[0];
+							const selectedText = selection.getSelectedText();
 
-					const bookmark = range.createBookmark();
+							const ranges = selection.getRanges();
 
-					linkElement = new CKEDITOR.dom.element('a');
+							const range = ranges[0];
```

Comment:
@markocikos, I needed to debug and see the result of `ranges` to understand what the first position was. Can We go with:

```js
const [rangeElement] = selection.getRanges();
```

this API is very weird 😂 

<img width="882" alt="Screen Shot 2021-08-12 at 20 29 52" src="https://user-images.githubusercontent.com/7663513/129282689-2d9313ab-7f99-48f7-bfbe-ecefa79ea4b2.png">

Reply:
@diegonvs API kinda makes sense, if you have in mind that the goal is to define position of selected area. They do this by breaking selected area into rectangles, that can be defined by `x`, `y`, `height` and `width`. In example below, `ranges` will have three elements in array:

![screen](https://user-images.githubusercontent.com/5435511/129340105-70096e9e-0012-4092-a701-19c074826dff.png)

So, to get to the start of selection, you go to start of the first rectangle. There is an alternative, probably cleaner way to to do this, using `selection.getStartElement()`, but this has some limitations and quirky behavior, see even in this PR the `startElement.getAscendant` bit.

In this PR, I didn't change the behavior of `LinkConfirm`. I only renamed it from `okButton` and moved it to `_initToolbarItems`, along with other custom toolbar items.

As for using 
```js
const [firstRange] = selection.getRanges();
```
yeah, that works nicely! :+1:  I'm not used to using destructuring in this way, but I think I will start to moving forward.

Reply:
Nevermind me, I was thinking of `range.getClientRect()` :sweat_smile:  That gives the useful array of rectangles. I am not sure what's the case with multiple `range`s. Maybe some parts of ckeditor can have multiselection?

Reply:
I was taking a look in what means bookmark and I saw nothing unless circular pointings betweens createBooksmarks is used to create a bookmark to use in moveBookmark but I don't know what it is. 

Seeing source code now I understand what bookmark is, From: https://github.com/ckeditor/ckeditor4/blob/4e64f672190a14d3c38aebd94ae4fb10c79e4545/core/dom/range.js#L708

** a bookmark object, which can be later used to restore the range by using the {@link #moveToBookmark} function.

😂